vm/actor/src/builtin/shared.rs

```diff
@@ -25,11 +26,50 @@ where
     )?;
     let addrs: MinerAddrs = ret.deserialize()?;
 
-    Ok((addrs.owner, addrs.worker))
+    Ok((addrs.owner, addrs.worker, addrs.control_addrs))
 }
 
 #[derive(Serialize_tuple, Deserialize_tuple)]
 struct MinerAddrs {
     owner: Address,
     worker: Address,
+    control_addrs: Vec<Address>,
+}
+
+/// ResolveToIDAddr resolves the given address to it's ID address form.
+/// If an ID address for the given address dosen't exist yet, it tries to create one by sending
+/// a zero balance to the given address.
+pub(crate) fn resolve_to_id_addr<BS, RT>(
+    rt: &mut RT,
+    address: &Address,
+) -> Result<Address, Box<dyn StdError>>
+where
+    BS: BlockStore,
+    RT: Runtime<BS>,
+{
+    // if we are able to resolve it to an ID address, return the resolved address
+    if let Some(addr) = rt.resolve_address(address)? {
+        return Ok(addr);
+    }
+
+    // send 0 balance to the account so an ID address for it is created and then try to resolve
+    rt.send(
+        *address,
+        METHOD_SEND,
+        Default::default(),
+        Default::default(),
+    )
+    .map_err(|e| {
+        e.wrap(&format!(
+            "failed to send zero balance to address {}",
+            address
+        ))
+    })?;
+
+    match rt.resolve_address(address)? {
+        Some(addr) => Ok(addr),
```

Comment:
Something like this might be better

match rt.resolve_address(address).ok_or_else(|| "failed to resolve address {} to ID address even after sending zero balance")?

Reply:
ah yeah, that was intentional, but yeah this is probably more idiomatic. (I also realized I forgot to use the `{}` to place in the address lmao)